Fix ::: Android ::: PickContacts when there are multiple numbers

OS-pedrogustavobilro opened the pull request:
## Description

The Pick Contacts feature was searching the contacts using a raw contact id instead of the actual contact id - which means if a contact has multiple raw contacts - e.g. multiple phone numbers added at different points in time - not all contact information was being returned.

This PR fixes that issue.

## Context

References: https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment

## Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply -->
- [X] Fix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Refactor (cosmetic changes)
- [ ] Breaking change (change that would cause existing functionality to not work as expected)

## Platforms affected
- [X] Android
- [ ] iOS
- [ ] JavaScript

## Tests

Tested the following in multiple Android versions:

- Creating a contact with multiple phone numbers (either via the Contacts Sample App or the phone's actual Contacts app) - Then using find and pick features - they should return the exact same results
- Confirming that saving contacts with same name a different phone still working.



## Checklist
<!--- Go over all the following items and put an `x` in all the boxes that apply -->
- [ ] Pull request title follows the format `RNMT-XXXX <title>`
- [ ] Code follows code style of this project
- [ ] CHANGELOG.md file is correctly updated
- [ ] Changes require an update to the documentation
	- [ ] Documentation has been updated accordingly

OS-pedrogustavobilro committed:
fix: Picking contact when there are multiple phone numbers / emails.

In other words, when there are multiple contact ID's.

References:
- https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment

OS-pedrogustavobilro committed:
fix: Saving contact using Raw Id.

References:
- https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment

alexgerardojacinto commented on src/android/ContactAccessorSdk5.java:
```diff
@@ -307,24 +306,41 @@ public JSONArray search(JSONArray fields, JSONObject options) {
-        return getContactById(id, null);
+        Cursor c = mApp.getActivity().getContentResolver().query(
+                ContactsContract.Data.CONTENT_URI,
+                null,
```
Not super important, but maybe we could explain why we pass `null` here.

OS-pedrogustavobilro replied:
This is how it was before, the `projection` field being passed is null to fetch all columns.

alexgerardojacinto replied:
Sorry, what I meant was to ask if we could add a comment in the code explaining why the `null`

OS-pedrogustavobilro replied:
Added comments in c0f30503dc28fc30b3f89e00be26af4fad66873c

alexgerardojacinto commented on src/android/ContactAccessorSdk5.java:
```diff
@@ -307,24 +306,41 @@ public JSONArray search(JSONArray fields, JSONObject options) {
+    public JSONObject getContactByRawId(String rawId) throws JSONException {
         Cursor c = mApp.getActivity().getContentResolver().query(
                 ContactsContract.Data.CONTENT_URI,
                 null,
```
same as above

alexgerardojacinto commented on src/android/ContactAccessorSdk5.java:
```diff
@@ -307,24 +306,41 @@ public JSONArray search(JSONArray fields, JSONObject options) {
         HashMap<String, Boolean> populate = buildPopulationSet(
-                new JSONObject().put("desiredFields", desiredFields)
-                );
+                new JSONObject().put("desiredFields", null)
```
same as above

OS-pedrogustavobilro replied:
We provide null here to create a population set with all attributes - as it was being done before - it wasn't changed here

The search for contacts functionality allows to specify in the javascript which fields do we want to return - this one does not. Although from my tests, the `desiredFields` in search comes as an empty array, so it has the same result.

alexgerardojacinto replied:
Sorry, what I meant was to ask if we could add a comment in the code explaining why the `null`

OS-ricardomoreirasilva commented on src/android/ContactAccessorSdk5.java:
```diff
@@ -307,24 +306,41 @@ public JSONArray search(JSONArray fields, JSONObject options) {
     public JSONObject getContactById(String id) throws JSONException {
-        // Call overloaded version with no desiredFields
-        return getContactById(id, null);
+        Cursor c = mApp.getActivity().getContentResolver().query(
+                ContactsContract.Data.CONTENT_URI,
+                null,
+                ContactsContract.Data.CONTACT_ID + " = ? ",
+                new String[] { id },
+                ContactsContract.Data.CONTACT_ID + " ASC"
+        );
+        return getContactById(c);
```
This code is a duplicate of `getContactByRawId` except for the `ContactsContract.Data` value. Couldn't we reuse it?

OS-pedrogustavobilro replied:
Yeah I suppose we could add the the row id instead as an argument to the common method and perform the query with it.

OS-pedrogustavobilro replied:
Done in 08b24f687c231415dfe9ce74257d808a7dd0ffe7

OS-pedrogustavobilro committed:
refactor: Re-use cursor query by passing the variable row id

References:
- https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment

OS-pedrogustavobilro committed:
docs: Comments on why null is passed

References:
- https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment

OS-ricardomoreirasilva commented on src/android/ContactAccessorSdk5.java:
```diff
@@ -307,24 +306,37 @@ public JSONArray search(JSONArray fields, JSONObject options) {
+        return getContactById(ContactsContract.Data.RAW_CONTACT_ID, rawId);
+    }
+
+    private JSONObject getContactById(String idKey, String idValue) throws JSONException {
```
I'm leaving my approval but as a note, since we're passing a key and a value, wouldn't it make sense to pass a single Pair-type argument?

OS-pedrogustavobilro replied:
In actuality, now that I'm seeing again, `idKey` is a bad name for it. It should be more like `idColumnName` and the other `idValue` or just `id`. It's not a key-value pair per se, since it isn't a map/dictionary, but rather a DB.

Does that make sense?

OS-pedrogustavobilro replied:
I ended up doing that change - 2ad91bd1b6a4bc3f4856fcd62ea94fa99a99a233 - if you'd like to take another look.

OS-ricardomoreirasilva replied:
It does. Still, even if the relationship is not a key-value pair per se, both arguments relate to each other. That's why I see value in grouping them. A structure/class for this specific use case seems too much so I suggested the Pair structure.

OS-pedrogustavobilro replied:
Understood, The `Pair` class is kotlin specific I think, but I'll use something similar in Java.

OS-pedrogustavobilro replied:
1b59b5055e9393760e707d2fc3070f5e09430025

OS-pedrogustavobilro committed:
refactor: Argument names

References:
- https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment

OS-pedrogustavobilro committed:
refactor: Pass key-value argument to common method

References:
- https://outsystemsrd.atlassian.net/wiki/spaces/RDME/pages/4426137659/iOS+18+Android+15+Assessment
